### PR TITLE
feat(orderby): added util for orderBy support in api calls

### PR DESCRIPTION
**Pull request by Arqu**

### CODE REVIEW

File: order.go [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+	// OrderASC defines the ascending order keyword
+	OrderASC = "asc"
+	// OrderDESC defines the descending order keyword
+	OrderDESC = "desc"
+)
+
+// OrderBy represents ordering information
+type OrderBy struct {
+	OrderItems []Order
+}
```
Comment (b5) [inline order.go]: love the `OrderBy` name. I think this type should be defined on the slice of items:

```go
// OrderBy represents ordering information
type OrderBy []Order
```

File: order.go [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+// NewOrderBy constructs a basic order by struct
+func NewOrderBy() OrderBy {
+	return OrderBy{
+		OrderItems: []Order{},
+	}
+}
+
+// NewOrder constructs a basic order struct
+func NewOrder(key string, orderDirection string) Order {
```
Comment (b5) [inline order.go]: https://www.youtube.com/watch?v=FYH8DsU2WCk

Also, group same-type declarations:
```go
func NewOrder(key, direction string) Order {
```

File: order.go [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+	Key       string
+	Direction string
+}
+
+// NewOrderBy constructs a basic order by struct
+func NewOrderBy() OrderBy {
+	return OrderBy{
+		OrderItems: []Order{},
+	}
+}
```
Comment (b5) [inline order.go]: Generally I prefer constructors appear just after the type declaration (this is where godoc puts them). In this case I think the constructor should be dropped if we switch to a type declaration on a slice, and direct users to use slice allocations to declare OrderBys:

```
apiutil.OrderBy{}
```

File: order.go [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+	return Order{
+		Key:       key,
+		Direction: orderDirection,
+	}
+}
+
+// AddOrdering allows to add items to the current ordering
+func (o *OrderBy) AddOrdering(order ...Order) {
+	o.OrderItems = append(o.OrderItems, order...)
+}
```
Comment (b5) [inline order.go]: if declared on a slice this gets easier:

```go
o := OrderBy{}
o = append(o, NewOrder("foo", "desc"))
```

File: order.go [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+// String implements the stringer interface for OrderBy
+func (o OrderBy) String() string {
+	var b bytes.Buffer
```
Comment (b5) [inline order.go]: use a [`strings.Builder`](https://golang.org/pkg/strings/#Builder) instead. It's more efficient, but the real benefit is dropping the need for the  `bytes` package from this file

File: order_test.go [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+	cases := []struct {
+		description string
+		query       string
+		expNumber   int
+		expOrderBy  string
+	}{
+		{"no orderBy specified", "", 0, ""},
+		{"orderBy with no direction (defaults to desc)", "title", 1, "title,desc"},
+		{"invalid orderBy", "title,asc,date", 0, ""},
+		{"orderBy with multiple parameters", "title,asc,date,desc", 2, "title,asc,date,desc"},
```
Comment (b5) [inline order_test.go]: what happens if I do `"title, date"`? Do I get `title,asc,date,asc`? Illegal?

File: order.go [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+}
+
+// OrderByFromRequest extracts orderBy params from an http request
+func OrderByFromRequest(r *http.Request) OrderBy {
+	orderBy := r.FormValue("orderBy")
+	return NewOrderByFromString(orderBy)
+}
+
+// NewOrderByFromString converts a commaa delimited string to an OrderBy struct
+func NewOrderByFromString(orderBy string) OrderBy {
```
Comment (b5) [inline order.go]: wondering if this should return an error on bad parsing. I'm sure you considered it. Thoughts?

File: order.go (hunk 2 above)
Comment (Arqu) [reply order.go]: Dang, haven't heard this one in quite a while!

File: order_test.go (hunk 6 above)
Comment (Arqu) [reply order_test.go]: It makes a key " date" which you can but don't have to process wherever this is used. I could go and start stripping weird signs but that's and endless thing. Maybe just whitespace front/back?

File: order.go (hunk 7 above)
Comment (Arqu) [reply order.go]: Compared to tha Page implementation, which has "defaults" passed in. 
~~Maybe we could do the same here?~~ However defaults here would mean assumptions about the keys that exist etc. 

Also the implementation guards against odd cases, direction is defaulted to `asc` and `desc` if no match is found. I gues the user is responsible for the end processing of the keys?

**Review by b5 — APPROVED** [review]
👍

**Pull request merged**